File: backend/app/application/services/settings_service.py
"""
Settings service.

Handles user settings management.
"""
from typing import Optional
from uuid import UUID

from app.domain.entities.settings import UserSettings
from app.domain.repositories.settings_repository import SettingsRepository


class SettingsService:
    """Service handling user settings operations."""

    def __init__(self, settings_repository: SettingsRepository):
        self._settings_repo = settings_repository

    async def get_user_settings(self, user_id: UUID) -> UserSettings:
        """
        Get settings for a user, creating defaults if needed.

        Args:
            user_id: User's ID.

        Returns:
            User's settings.
        """
        settings = await self._settings_repo.get_by_user_id(user_id)

        if not settings:
            settings = await self._settings_repo.create_default(user_id)

        return settings

    async def update_settings(
        self,
        user_id: UUID,
        sound_effects: Optional[bool] = None,
        notifications: Optional[bool] = None,
        show_wpm_live: Optional[bool] = None,
        show_accuracy_live: Optional[bool] = None,
        theme: Optional[str] = None,
        keyboard_layout: Optional[str] = None,
    ) -> UserSettings:
        """
        Update user settings.

        Args:
            user_id: User's ID.
            sound_effects: Enable/disable sound effects.
            notifications: Enable/disable notifications.
            show_wpm_live: Show live WPM.
            show_accuracy_live: Show live accuracy.
            theme: UI theme.
            keyboard_layout: Keyboard layout.

        Returns:
            Updated settings.
        """
        settings = await self.get_user_settings(user_id)

        if sound_effects is not None:
            settings.sound_effects = sound_effects
        if notifications is not None:
            settings.notifications = notifications
        if show_wpm_live is not None:
            settings.show_wpm_live = show_wpm_live
        if show_accuracy_live is not None:
            settings.show_accuracy_live = show_accuracy_live
        if theme is not None:
            settings.theme = theme
        if keyboard_layout is not None:
            settings.keyboard_layout = keyboard_layout

        return await self._settings_repo.update(settings)

File: frontend/src/services/settingsService.ts
/**
 * Settings Service
 */
import { apiClient } from './api';

export interface UserSettings {
  soundEffects: boolean;
  notifications: boolean;
  showWpmLive: boolean;
  showAccuracyLive: boolean;
  theme: string;
  keyboardLayout: string;
}

interface SettingsResponse {
  sound_effects: boolean;
  notifications: boolean;
  show_wpm_live: boolean;
  show_accuracy_live: boolean;
  theme: string;
  keyboard_layout: string;
}

function mapSettingsResponse(response: SettingsResponse): UserSettings {
  return {
    soundEffects: response.sound_effects,
    notifications: response.notifications,
    showWpmLive: response.show_wpm_live,
    showAccuracyLive: response.show_accuracy_live,
    theme: response.theme,
    keyboardLayout: response.keyboard_layout,
  };
}

export const settingsService = {
  async getSettings(): Promise<UserSettings> {
    const response = await apiClient.get<SettingsResponse>('/settings');
    return mapSettingsResponse(response);
  },

  async updateSettings(settings: Partial<UserSettings>): Promise<UserSettings> {
    const request: Record<string, unknown> = {};
    if (settings.soundEffects !== undefined) request.sound_effects = settings.soundEffects;
    if (settings.notifications !== undefined) request.notifications = settings.notifications;
    if (settings.showWpmLive !== undefined) request.show_wpm_live = settings.showWpmLive;
    if (settings.showAccuracyLive !== undefined) request.show_accuracy_live = settings.showAccuracyLive;
    if (settings.theme !== undefined) request.theme = settings.theme;
    if (settings.keyboardLayout !== undefined) request.keyboard_layout = settings.keyboardLayout;

    const response = await apiClient.put<SettingsResponse>('/settings', request);
    return mapSettingsResponse(response);
  },
};
